File: src/port_doggie.erl
%% port_doggie.erl
%- Erlang version of port_doggie.py
% 
%- start port_slippers in a separate shell (python port_slippers.py)
%- start erl
%  c(port_doggie).
%- save yourself some typing:
%  PD = port_doggie.
%  PD:start().
%- get some meter data:
%  PD:m().
%- get some spec data:
%  PD:s().
%- ...lather, rinse, repeat...
%- shut down:
%  PD:q().

-module(port_doggie).
-export([start/0,start/3,s/0,s/1,m/0,m/1,q/0]).

-define(CMD_PORT,19001).
-define(SPEC_PORT,19002).
-define(METER_PORT,19003).
-define(DFLT_SPEC_LABEL,4277009102).
-define(DFLT_METER_LABEL,3735928559).
-define(TIMEOUT,1000).

start() ->
    io:format("start()~n",[]),
    start(?CMD_PORT,?SPEC_PORT,?METER_PORT).

start(CmdPort,SpecPort,MeterPort) ->
    CPid = spawn(
	     fun() -> server("localhost",CmdPort,SpecPort,MeterPort) end
	    ),
    io:format("server is ~p~n",[CPid]),
    register(cmdr, CPid).

%%----------------------------------------
%% pass along requests for spectrum, meter
%%----------------------------------------

server(ToHost,ToPort,SpecPort,MeterPort) ->
    {ok, CmdSock} = gen_udp:open(0,[binary]),
    io:format("server socket is ~p~n",[CmdSock]),
    % here's why this is a silly example:
    {ok,SpecSock} = gen_udp:open(SpecPort,[binary,{recbuf,32768}]),
    io:format("spec sucker socket is ~p~n",[SpecSock]),
    {ok,MeterSock} = gen_udp:open(MeterPort,[binary]),
    io:format("meter sucker socket is ~p~n",[MeterSock]),
    loop(CmdSock,ToHost,ToPort,SpecSock,MeterSock).

%%-
%% send a command string, wait for response
%% expects an Erlang string (list), converts to packed binary
%%-
ask(Req,Sock,Host,Port) ->
    io:format("ask sending ~p to ~p,~p~n",[Req,Host,Port]),
    Pkt = list_to_binary(Req),
    gen_udp:send(Sock,Host,Port,[Pkt]),
    receive
    	{udp,_Serv,_Addr,_Port,<<"ok">>} ->
	    io:format("ask gets ok from server~n",[]);
	{udp,_Serv,_Addr,_Port,<<"error">>} ->
	    io:format("ask gets error from server~n",[]);
	Other ->
	    io:format("ask sees Other ~p~n",[Other])
    end.

%%-
%% absorb binary payload of a Type
%%-
suck(Sock,Type) ->
    io:format("suck ~p from ~p~n",[Type,Sock]),
    receive
	{udp,Sock,Host,Port,Pkt} ->
	    io:format("~p ~p ~p ~p~n",[Type,Sock,Host,Port]),
	    {Label,Vals} = brkout(Pkt),
	    io:format("~p ~p ~p~n",[Type,Label,Vals]);
	Other ->
	    io:format("~p garbage ~p~n",[Type,Other])
    after
	?TIMEOUT ->
	    io:format("suck times out~n",[])
    end.

%%-
%% main command loop
%%-
loop(CmdSock,CmdHost,CmdPort,SpecSock,MeterSock) ->
    receive
	{From,spectrum,Label} ->
	    io:format("loop sees spectrum req from ~p, label ~p~n",[From,Label]),
	    Cmd = "reqSpectrum " ++ integer_to_list(Label) ++ "\n",
	    ask(Cmd,CmdSock,CmdHost,CmdPort),
	    suck(SpecSock,"spec"),
	    From ! {self(),spectrum,ok},
	    loop(CmdSock,CmdHost,CmdPort,SpecSock,MeterSock);

	{From,meter,Label} ->
	    io:format("loop sees meter req from ~p, label ~p~n",[From,Label]),
	    Cmd = "reqMeter " ++ integer_to_list(Label) ++ "\n",
	    ask(Cmd,CmdSock,CmdHost,CmdPort),
	    suck(MeterSock,"meter"),
	    From ! {self(),meter,ok},
	    loop(CmdSock,CmdHost,CmdPort,SpecSock,MeterSock);

	{From,stop} ->
	    io:format("loop sees stop req by ~p~n",[From])
    end.

%%---------------------------------
%% "local" user interface functions
%%---------------------------------

%%-
%% initiate request for spectrum
%%-

s() ->
    io:format("s()~n",[]),
    s(?DFLT_SPEC_LABEL).

s(Label) ->
    io:format("s(~p)~n",[Label]),
    cmdr ! {self(),spectrum,Label},
    receive
	{_,spectrum,ok} ->
	    io:format("s() received ack~n",[]);
	_ ->
	    io:format("s() failed to receive ack~n",[])
    end.

%%-
%% initiate request for meter
%%-

m() ->
    io:format("m()~n",[]),
    m(?DFLT_METER_LABEL).

m(Label) ->
    io:format("m(~p)~n",[Label]),
    cmdr ! {self(),meter,Label},
    receive
	{_,meter,ok} ->
	    io:format("m() received ack~n",[]);
	_ ->
	    io:format("m() failed to receive ack~n",[])
    end.

%%-
%% shut down
%%-
q() ->
    io:format("q()~n",[]),
    cmdr ! {self(),stop}.

%%-
%% extract 32-bit label and float vector from binary payload
%%-
brkout(Bin) ->
    <<Label:32,Rest/binary>> = Bin,
    {Label,brkoutf(Rest)}.

%%-
%% brkoutf
%% convert single binary containing 32-bit packed floats to list
%%-

brkoutf(Bin) ->
    brkoutf(Bin,[]).
brkoutf(Bin,Acc) ->
    <<V:32/float,Rest/binary>> = Bin,
    Vals = [V|Acc],
    if
	size(Rest) > 0 ->
	    brkoutf(Rest,Vals);
	true ->
	    lists:reverse(Vals)
    end.

File: src/port_doggie.py
#! /usr/bin/env python
"""
A pair of classes to exercise the spectrum and meter data capture
from an sdr-core. Does fetch on spectrum and meter ports of a running
process.

sdog = spec_doggie([host], [command-port], [spectrum-port])
then
label, stamp, spectrum = sdog.get()

mdog = meter_doggie([host], [command-port], [spectrum-port])
then
label, readings = meter_doggie([host], [command-port], [meter-port])

Standalone usage:

port_doggie [host cmd-port spec-data-port meter-data-port]

Reads lines from stdin
s	means read spectrum
m	means read meter

Default host		localhost
	cmd-port	19001
        spec-data-port	19002
        meter-data-port	19003
"""

import sys
import socket
import select
import numpy as N

CMD_PORT = 19001
SPEC_PORT = 19002
METER_PORT = 19003

SPEC_BUFSIZE = 32768
METER_BUFSIZE = 1024
DUMMY_BUFSIZE = 512

#------------------------------------------------------------------
# base doggie
#------------------------------------------------------------------

class doggie(object):
    """
    A requester for a packet of data from server on (host, port)
    Command issued and ack'ed on cmd_port
    Data returned on data_port
    Base case, simply do command and ignore data
    """

    def __init__(self, host, cmd_port, data_port):
        self.cmd_host = host
        self.cmd_port = cmd_port
        try:
            self.cmd_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        except Exception, e:
            print >>sys.stderr, "Can't open UDP command socket on %r, %r" % (host, cmd_port)
            print >>sys.stderr, type(e)
            sys.exit(1)
        self.cmd_sock.bind(('', 0))
        self.cmd_addr = self.cmd_host, self.cmd_port

        self.data_host = host
        self.data_port = data_port
        try:
            self.data_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        except Exception, e:
            print >>sys.stderr, "Can't open UDP data socket on %r, %r" % (host, data_port)
            print >>sys.stderr, type(e)
            sys.exit(1)
        self.data_sock.bind(('', self.data_port))

    def done(self):
        self.cmd_sock.close()
        self.data_sock.close()

    def _req(self, command):
        """
        Issue command, wait a little for response, check ack
        """
        # make request
        self.reqmsg = command
        self.cmd_sock.sendto(self.reqmsg, self.cmd_addr)
        # wait a bit for ack
        irdy, ordy, erdy = select.select([self.cmd_sock], [], [], 1.0)
        if not self.cmd_sock in irdy:
            print >>sys.stderr, "Late or missing ACK from command"
            return False
        self.ack, fromaddr = self.cmd_sock.recvfrom(DUMMY_BUFSIZE)
        if self.ack <> 'ok':
            print >>sys.stderr, "Command failed"
            return False
        return True

    def _inp(self, timeout):
        """
        Wait a bit for data, then read back
        """
        irdy, ordy, erdy = select.select([self.data_sock], [], [], timeout)
        if not self.data_sock in irdy:
            print >>sys.stderr, "Data late or missing"
            return False
        self.payload, self.fromaddr = self.data_sock.recvfrom(SPEC_BUFSIZE)
        return True

    def get(self, command):
        if not self._req(command):
            print >>sys.stderr, command, 'failed'
        return None


#------------------------------------------------------------------
# derived spectrum doggie
#------------------------------------------------------------------

class spec_doggie(doggie):
    """
    Spectrum fetcher
    """

    def __init__(self,
                 host      = 'localhost',
                 cmd_port  = CMD_PORT,
                 data_port = SPEC_PORT):
        super(spec_doggie, self).__init__(host, cmd_port, data_port)

    def get(self, token = 4277009102):
        self.token = token

        if not self._req('reqSpectrum' + ' '+ str(token) + '\n'):
            print >>sys.stderr, "Command failed"
            return None, None, None

        if not self._inp(1.0):
            print >>sys.stderr, "Spectrum data late or missing"
            return None, None, None
            
        # break out
        self.labelstr = self.payload[0:4]
        self.stampstr = self.payload[4:8]
        self.specpstr = self.payload[8: ]
        label = N.fromstring(self.labelstr, N.uint32)[0]
        stamp = N.fromstring(self.stampstr, N.uint32)[0]
        specp = N.fromstring(self.specpstr, N.float)
        return label, stamp, specp


#------------------------------------------------------------------
# derived meter doggie
#------------------------------------------------------------------

class meter_doggie(doggie):
    """
    Meter fetcher
    """

    def __init__(self,
                 host      = 'localhost',
                 cmd_port  = CMD_PORT,
                 data_port = METER_PORT):
        super(meter_doggie, self).__init__(host, cmd_port, data_port)

    def get(self, token = 3735928559):
        self.token = token

        if not self._req('reqMeter' + ' ' + str(token) + '\n'):
            print >>sys.stderr, "Command failed"
            return None, None

        if not self._inp(1.0):
            print >>sys.stderr, "Spectrum data late or missing"
            return None, None
            
        # break out
        self.labelstr = self.payload[0:4]
        self.mvalsstr = self.payload[4: ]
        label = N.fromstring(self.labelstr, N.uint32)[0]
        mvals = N.fromstring(self.mvalsstr, N.float)
        return label, mvals


#------------------------------------------------------------------
#------------------------------------------------------------------

if __name__ == '__main__':
    if len(sys.argv) == 1:
        host = 'localhost'
        cport = CMD_PORT
        sport = SPEC_PORT
        mport = METER_PORT
    elif len(sys.argv) == 5:
        host = sys.argv[1]
        cport = int(sys.argv[2])
        sport = int(sys.argv[3])
        mport = int(sys.argv[4])
    else:
        print >>sys.stderr, __doc__
        sys.exit(2)

    sdog = spec_doggie(host, cport, sport)
    print >>sys.stderr, "spec doggie is ready"
    mdog = meter_doggie(host, cport, mport)
    print >>sys.stderr, "meter doggie is ready"

    while True:
        try:
            line = sys.stdin.readline()
        except KeyboardInterrupt:
            break;
        if not line:
            break
        if line[0] == 's':
            print 'Fetch spectrum'
            pkt = sdog.get()
            print pkt
#            label, stamp, specp = sdog.get()
#            print 'Got', label, stamp, specp
        elif line[0] == 'm':
            print 'Fetch meter'
            pkt = mdog.get()
            print pkt
#            label, mvals = mdog.get()
#            print 'Got', label, mvals
        else:
            print 'Can\'t do that'

    sdog.done()
    mdog.done()
